apps/beeswax/src/beeswax/server/dbms.py

```diff
@@ -83,7 +88,57 @@ def get_query_server_config(name='beeswax', server=None, cluster=None):
-
+  if name == "llap":
+    activeEndpoint = cache.get('llap')
+    if activeEndpoint is None:
+      if HIVE_DISCOVERY_LLAP.get() == True:
+        LOG.debug("Checking zookeeper for Hive Server Interactive endpoint")
+        from kazoo.client import KazooClient
+        zk = KazooClient(hosts=libzookeeper_conf.ENSEMBLE.get(), read_only=True)
+        zk.start()
+        if HIVE_DISCOVERY_LLAP_HA.get() == True:
+          znode = "{0}/instances".format(HIVE_DISCOVERY_LLAP_ZNODE.get())
+          LOG.debug("Setting up LLAP with the following node {0}".format(znode))
+          if zk.exists(znode):
+            hiveservers = zk.get_children(znode)
+            for server in hiveservers:
+              llap_servers= json.loads(zk.get("{0}/{1}".format(znode, server))[0])["internal"][0]
+              if llap_servers["api"] == "activeEndpoint":
+                cache.set("llap", json.dumps({"host": llap_servers["addresses"][0]["host"], "port": llap_servers["addresses"][0]["port"]}), CACHE_TIMEOUT.get())
+          else:
+            LOG.error("LLAP Endpoint not found, reverting to HiveServer2")
+            cache.set("llap", json.dumps({"host": HIVE_SERVER_HOST.get(), "port": hive_site.hiveserver2_thrift_http_port()}), CACHE_TIMEOUT.get())
+        else:
+          znode = "{0}".format(HIVE_DISCOVERY_LLAP_ZNODE.get())
+          LOG.debug("Setting up LLAP with the following node {0}".format(znode))
+          if zk.exists(znode):
+            hiveservers = zk.get_children(znode)
+            for server in hiveservers:
+              cache.set("llap", json.dumps({"host": server.split(';')[0].split('=')[1].split(":")[0], "port": server.split(';')[0].split('=')[1].split(":")[1]}))
+        zk.stop()
+      else:
+        LOG.debut("Zookeeper Discovery not enabled, reverting to config values")
+        cache.set("llap", json.dumps({"host": LLAP_SERVER_HOST.get(), "port": LLAP_SERVER_PORT.get()}), CACHE_TIMEOUT.get())
+    activeEndpoint = json.loads(cache.get("llap"))
+  elif name != 'hms' and name != 'impala':
+    activeEndpoint = cache.get("hiveserver2")
+    if activeEndpoint is None:
+      if HIVE_DISCOVERY_HS2.get() == True:
+        from kazoo.client import KazooClient
+        zk = KazooClient(hosts=libzookeeper_conf.ENSEMBLE.get(), read_only=True)
+        zk.start()
+        znode = HIVE_DISCOVERY_HIVESERVER2_ZNODE.get()
+        LOG.info("Setting up Hive with the following node {0}".format(znode))
+        if zk.exists(znode):
+          hiveservers = zk.get_children(znode)
+          server_to_use = 0 # if CONF.HIVE_SPREAD.get() randint(0, len(hiveservers)-1) else 0
+          cache.set("hiveserver2", json.dumps({"host": hiveservers[server_to_use].split(";")[0].split("=")[1].split(":")[0], "port": hiveservers[server_to_use].split(";")[0].split("=")[1].split(":")[1]}))
+        else:
+          cache.set("hiveserver2", json.dumps({"host": HIVE_SERVER_HOST.get(), "port": hive_site.hiveserver2_thrift_http_port()}))
```

Comment:
We lose the 'if not cluster_config else cluster_config.get('server_host')' probably here
(but might be fine as I am currently cleaning this up)

Reply:
not sure if i have a task here

Reply:
No, was more a note for myself!